README


This archive contains the distribution ORDB-CPAN-Mandriva, version
1.100230:

  an ORM for CPAN packages available in Mandriva

This software is copyright (c) 2010 by Jerome Quelin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for ORDB-CPAN-Mandriva

1.100230  2010-01-23 16:32:40 Europe/Paris
 - initial version


lib/ORDB/CPAN/Mandriva.pm
# 
# This file is part of ORDB-CPAN-Mandriva
# 
# This software is copyright (c) 2010 by Jerome Quelin.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use 5.008005;
use strict;
use warnings;

package ORDB::CPAN::Mandriva;
our $VERSION = '1.100230';
# ABSTRACT: an ORM for CPAN packages available in Mandriva

use ORLite::Mirror 1.12 ();

sub import {
    my $class = shift;

    # prevent double-initialisation
    $class->can('orlite') or
    ORLite::Mirror->import( {
        url    => 'http://ryu.zarb.org/~jquelin/cpan_Mandriva.db',
        maxage => 7 * 24 * 60 * 60,
    } );

    return 1;
}



# -- methods autogenerated by orlite


1;


=pod

=head1 NAME

ORDB::CPAN::Mandriva - an ORM for CPAN packages available in Mandriva

=head1 VERSION

version 1.100230

=head1 SYNOPSIS

    use ORDB::CPAN::Mandriva;
    my $nbmodules = ORDB::CPAN::Mandriva::Module->count;
    my $cpandists = ORDB::CPAN::Mandriva->selectcol_arrayref(
        'SELECT DISTINCT dist FROM module ORDER BY dist'
    );

=head1 DESCRIPTION

This module is an easy way to fetch a database listing all Perl modules
& distributions packaged within Mandriva Linux distribution.

When using it, it will automatically & silently download it from the
original source and copy it in a local directory, letting you focus on
what you want with the data itself.

Check the F<examples> directory for some ideas on how to use it.

=head1 METHODS

=head2 my $string = ORDB::CPAN::Mandriva->dsn;

The C<dsn> accessor returns the dbi connection string used to connect to
the SQLite database as a string.

=head2 my $handle = ORDB::CPAN::Mandriva->dbh;

To reliably prevent potential SQLite deadlocks resulting from multiple
connections in a single process, each L<ORLite> package will only ever
maintain a single connection to the database.

During a transaction, this will be the same (cached) database handle.

Although in most situations you should not need a direct DBI connection
handle, the C<dbh> method provides a method for getting a direct
connection in a way that is compatible with ORLite's connection
management.

Please note that these connections should be short-lived, you should
never hold onto a connection beyond the immediate scope.

The transaction system in L<ORLite> is specifically designed so that
code using the database should never have to know whether or not it is
in a transation.

Because of this, you should B<never> call the -E<gt>disconnect method on
the database handles yourself, as the handle may be that of a currently
running transaction.

Further, you should do your own transaction management on a handle
provided by the <dbh> method.

In cases where there are extreme needs, and you B<absolutely> have to
violate these connection handling rules, you should create your own
completely manual DBI-E<gt>connect call to the database, using the
connect string provided by the C<dsn> method.

The C<dbh> method returns a L<DBI::db> object, or throws an exception on
error.

=head2 ORDB::CPAN::Mandriva->begin;

The C<begin> method indicates the start of a transaction.

In the same way that L<ORLite> allows only a single connection, likewise
it allows only a single application-wide transaction.

No indication is given as to whether you are currently in a transaction
or not, all code should be written neutrally so that it works either way
or doesn't need to care.

Returns true or throws an exception on error.

=head2 ORDB::CPAN::Mandriva->commit;

The C<commit> method commits the current transaction. If called outside
of a current transaction, it is accepted and treated as a null
operation.

Once the commit has been completed, the database connection falls back
into auto-commit state. If you wish to immediately start another
transaction, you will need to issue a separate -E<gt>begin call.

Returns true or throws an exception on error.

=head2 ORDB::CPAN::Mandriva->rollback;

The C<rollback> method rolls back the current transaction. If called outside
of a current transaction, it is accepted and treated as a null operation.

Once the rollback has been completed, the database connection falls back
into auto-commit state. If you wish to immediately start another
transaction, you will need to issue a separate -E<gt>begin call.

If a transaction exists at END-time as the process exits, it will be
automatically rolled back.

Returns true or throws an exception on error.

=head2 ORDB::CPAN::Mandriva->do(...);

    ORDB::CPAN::Mandriva->do(
        insert into table (foo, bar) values (?, ?)', {},
        $foo_value,
        $bar_value,
    );

The C<do> method is a direct wrapper around the equivalent L<DBI>
method, but applied to the appropriate locally-provided connection or
transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectall_arrayref(...);

The C<selectall_arrayref> method is a direct wrapper around the
equivalent L<DBI> method, but applied to the appropriate locally-
provided connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectall_hashref(...);

The C<selectall_hashref> method is a direct wrapper around the
equivalent L<DBI> method, but applied to the appropriate locally-
provided connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectcol_arrayref(...);

The C<selectcol_arrayref> method is a direct wrapper around the
equivalent L<DBI> method, but applied to the appropriate locally-
provided connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectrow_array(...);

The C<selectrow_array> method is a direct wrapper around the equivalent
L<DBI> method, but applied to the appropriate locally-provided
connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectrow_arrayref(...);

The C<selectrow_arrayref> method is a direct wrapper around the
equivalent L<DBI> method, but applied to the appropriate locally-
provided connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->selectrow_hashref(...);

The C<selectrow_hashref> method is a direct wrapper around the
equivalent L<DBI> method, but applied to the appropriate locally-
provided connection or transaction.

It takes the same parameters and has the same return values and error
behaviour.

=head2 ORDB::CPAN::Mandriva->prepare(...);

The C<prepare> method is a direct wrapper around the equivalent L<DBI>
method, but applied to the appropriate locally-provided connection or
transaction

It takes the same parameters and has the same return values and error
behaviour.

In general though, you should try to avoid the use of your own prepared
statements if possible, although this is only a recommendation and by no
means prohibited.

=head2 ORDB::CPAN::Mandriva->pragma(...);

  # Get the user_version for the schema
  my $version = ORDB::CPAN::Mandriva->pragma('user_version');

The C<pragma> method provides a convenient method for fetching a pragma
for a datase. See the SQLite documentation for more details.

=head1 SEE ALSO

You can find more information on this module at:

=over 4

=item * Search CPAN

L<http://search.cpan.org/dist/ORDB-CPAN-Mandriva>

=item * See open / report bugs

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=ORDB-CPAN-Mandriva>

=item * Git repository

L<http://github.com/jquelin/ordb-cpan-mandriva>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/ORDB-CPAN-Mandriva>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/ORDB-CPAN-Mandriva>

=back

=head1 AUTHOR

  Jerome Quelin

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Jerome Quelin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



lib/ORDB/CPAN/Mandriva/Module.pm
# 
# This file is part of ORDB-CPAN-Mandriva
# 
# This software is copyright (c) 2010 by Jerome Quelin.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
use 5.008005;
use strict;
use warnings;

package ORDB::CPAN::Mandriva::Module;
our $VERSION = '1.100230';
# ABSTRACT: orlite for module table in database

# -- attributes



# -- methods autogenerated by orlite


1;


=pod

=head1 NAME

ORDB::CPAN::Mandriva::Module - orlite for module table in database

=head1 VERSION

version 1.100230

=head1 DESCRIPTION

This class models the C<module> table in the database. It can be used
either with class methods, or as objects mapping one row of the table.

=head1 ATTRIBUTES

=head2 my $str = $module->module;

The name of the module, eg L<ORDB::CPAN::Mandriva>.

=head2 my $str = $module->version;

The version of the module (not the dist, nor the package). It may be
null.

=head2 my $str = $module->dist;

This is the CPAN distribution the module is part of, eg
C<ORDB-CPAN-Mandriva>. It may be null.

=head2 my $str = $module->pkgname;

This is the name of the package holding this module in Mandriva Linux
distribution. Chances are that it looks like C<perl-ORDB-CPAN-Mandriva>.

=head1 METHODS

=head2 select

    # Get all objects in list context
    my @list = ORDB::CPAN::Mandriva::Module->select;

    # Get a subset of objects in scalar context
    my $array_ref = ORDB::CPAN::Mandriva::Module->select(
        'where pkgname = ? order by module',
        'perl-ORDB-CPAN-Mandriva',
    );

The C<select> method executes a typical SQL C<SELECT> query on the
module table.

It takes an optional argument of a SQL phrase to be added after the
C<FROM module> section of the query, followed by variables to be bound
to the placeholders in the SQL phrase. Any SQL that is compatible with
SQLite can be used in the parameter.

Returns a list of L<ORDB::CPAN::Mandriva::Module> objects when called in
list context, or a reference to an ARRAY of
L<ORDB::CPAN::Mandriva::Module> objects when called in scalar context.

Throws an exception on error, typically directly from the L<DBI> layer.

=head2 count

    # How many objects are in the table
    my $rows = ORDB::CPAN::Mandriva::Module->count;

    # How many objects
    my $small = ORDB::CPAN::Mandriva::Module->count(
        'where pkgname = ?',
        'perl-ORDB-CPAN-Mandriva',
    );

The C<count> method executes a C<SELECT COUNT(*)> query on the
module table.

It takes an optional argument of a SQL phrase to be added after the
C<FROM module> section of the query, followed by variables to be bound
to the placeholders in the SQL phrase. Any SQL that is compatible with
SQLite can be used in the parameter.

Returns the number of objects that match the condition.

Throws an exception on error, typically directly from the L<DBI> layer.

=head2 my $module = Module->new(%params);

The C<new> constructor is used to create a new abstract object that
is not (yet) written to the database.

Returns a new L<ORDB::CPAN::Mandriva::Module> object.

=head2 $module->insert;

The C<insert> method commits a new object (created with the C<new>
method) into the database.

Returns the object itself as a convenience, or throws an exception on
error, typically from the L<DBI> layer.

=head2 delete

    # Delete a single instantiated object
    $object->delete;

    # Delete multiple rows from the module table
    ORDB::CPAN::Mandriva::Module->delete(
        'where pkgname = ?',
        'perl-ORDB-CPAN-Mandriva'
    );

The C<delete> method can be used in a class form and an instance form.

When used on an existing L<ORDB::CPAN::Mandriva::Module> instance,
the C<delete> method removes that specific instance from the
C<module>, leaving the object intact for you to deal with post-delete
actions as you wish.

When used as a class method, it takes a compulsory argument of a SQL
phrase to be added after the C<DELETE FROM module> section of the query,
followed by variables to be bound to the placeholders in the SQL phrase.
Any SQL that is compatible with SQLite can be used in the parameter.

Returns true on success or throws an exception on error, or if you
attempt to call delete without a SQL condition phrase.

=head2 truncate

    # Delete all records in the module table
    ORDB::CPAN::Mandriva::Module->truncate;

To prevent the common and extremely dangerous error case where deletion
is called accidentally without providing a condition, the use of the
C<delete> method without a specific condition is forbidden.

Instead, the distinct method C<truncate> is provided to delete all
records in a table with specific intent.

Returns true, or throws an exception on error.

=head2 my $module = Module->create(%params);

The C<create> constructor is a one-step combination of C<new> and
C<insert> that takes the column parameters, creates a new
L<ORDB::CPAN::Mandriva::Module> object, inserts the appropriate row into
the C<module> table, and then returns the object.

Returns a new L<Module> object, or throws an exception on error,
typically from the L<DBI> layer.

=head1 AUTHOR

  Jerome Quelin

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Jerome Quelin.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;

my %module_build_args = (
                       'dist_abstract' => 'an ORM for CPAN packages available in Mandriva',
                       'dist_author' => [
                                          'Jerome Quelin'
                                        ],
                       'dist_name' => 'ORDB-CPAN-Mandriva',
                       'script_files' => [],
                       'module_name' => 'ORDB::CPAN::Mandriva',
                       'requires' => {
                                       'perl' => '5.008005',
                                       'Test::More' => 0,
                                       'English' => 0,
                                       'File::Find' => 0,
                                       'File::Temp' => 0,
                                       'ORLite::Mirror' => '1.12'
                                     },
                       'license' => 'perl'
                     );


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
